common/src/main/java/com/microsoft/identity/common/internal/telemetry/adapter/TelemetryAggregationAdapter.java

```diff
@@ -0,0 +1,115 @@
+import com.microsoft.identity.common.internal.util.StringUtil;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static com.microsoft.identity.common.internal.telemetry.TelemetryEventStrings.TELEMETRY_EVENT_API_END_EVENT;
+import static com.microsoft.identity.common.internal.telemetry.TelemetryEventStrings.TELEMETRY_EVENT_API_START_EVENT;
+import static com.microsoft.identity.common.internal.telemetry.TelemetryEventStrings.TELEMETRY_KEY_END_TIME;
+import static com.microsoft.identity.common.internal.telemetry.TelemetryEventStrings.TELEMETRY_KEY_EVENT_NAME;
+import static com.microsoft.identity.common.internal.telemetry.TelemetryEventStrings.TELEMETRY_KEY_EVENT_TYPE;
+import static com.microsoft.identity.common.internal.telemetry.TelemetryEventStrings.TELEMETRY_KEY_OCCUR_TIME;
+import static com.microsoft.identity.common.internal.telemetry.TelemetryEventStrings.TELEMETRY_KEY_START_TIME;
+
+public final class TelemetryAggregationAdapter implements ITelemetryAdapter<List<Map<String, String>>> {
+    private ITelemetryAggregatedObserver mObserver;
+
+    public TelemetryAggregationAdapter(@NonNull final ITelemetryAggregatedObserver observer) {
+        mObserver = observer;
+    }
+
+    public ITelemetryAggregatedObserver getObserver() {
+        return mObserver;
+    }
+
+    public void process(List<Map<String, String>> rawData) {
+        final Map<String, String> aggregatedData = new HashMap<>();
+        long apiStartTime = -1;
+        long apiEndTime = -1;
+
+        final Iterator<Map<String, String>> iterator = rawData.iterator();
+        while (iterator.hasNext()) {
+            Map<String, String> event = iterator.next();
+            final String eventName = event.get(TELEMETRY_KEY_EVENT_NAME);
+            final String eventType = event.get(TELEMETRY_KEY_EVENT_TYPE);
+
+            if (StringUtil.isEmpty(eventName)) {
+                aggregatedData.putAll(applyAggregationRule(event));
+                continue;
+            }
+
+            //only count the starting of event.
+            if (eventName.contains("start")) {
+                final String eventTypeCount = eventType + "_count";
+                aggregatedData.put(
+                        eventTypeCount,
+                        null == aggregatedData.get(eventTypeCount) ?
+                                "1"
+                                : String.valueOf(Integer.parseInt(aggregatedData.get(eventTypeCount)) + 1)
+                );
+            }
+
+            final long eventOccurTime = Long.parseLong(event.get(TELEMETRY_KEY_OCCUR_TIME));
+            if (eventName.equalsIgnoreCase(TELEMETRY_EVENT_API_START_EVENT)
+                    && (apiStartTime == -1
+                    || eventOccurTime < apiStartTime)) {
+                apiStartTime = eventOccurTime;
+            }
+
+            if (eventName.equalsIgnoreCase(TELEMETRY_EVENT_API_END_EVENT)
+                    && (apiEndTime == -1
```

Comment:
we probably won't need this -1 check (as eventOccurTime should always be bigger than -1).